### PR TITLE
Implement null propagation for JS.

### LINKED ISSUE #466: mongodb: runtime errors when null and undefined values occur in JavaScript
When we generate JavaScript (for `where` selectors and map-reduce ops), we don't currently guard against dereferencing `null`, `undefined`, or non-object values.

For example:
```sql
select * from zips where length(foo.bar) > 0
```

### LINKED ISSUE #528: mongodb: `$where` clause generates bad JS
When we generate the JS output that can be cut/pasted to Mongo, we do the wrong thing with `$where` filters:
```js
db.zips.mapReduce(
  function () { … },
  function (key, values) { … },
  { "query" : { "$where" : { "$code" : "…"}}})
```
There should be no `$code` field, it should be only `{ "$where" : "…"}`.

The current version fails with a “can’t canonicalize query” error.

### PR DESCRIPTION
**Pull request by sellout**
Fixes #466.

Mongo returns undefined (no field) for `foo.bar.baz` if any
sub-projection is undefined, but any expression other than projection
will result in null (EG, `foo.bar.baz + 4`). We now get those semantics
in JS as well.

JsCore does the right thing automatically, but also created a few helper
functions for JS until everything has been converted to JsCore.

* change $where to take a String, not Code (fixes #528),
* insert a space between unary op and argument (`typeof` is an UnOp),
* tightened up various conditions (`===` instead of `==`)
* add an integration test

The one inconsistency (maybe) is the handling of expressions that are
represented as properties in JS. EG, `length(foo)` would result in null
if it were handled by MongoDB, but because it becomes a
projection (`foo.length`) in JS, it ends up as undefined.

### CODE REVIEW

**Comment by sellout** [comment]
Also fixes #466, which is the whole point.

File: core/src/main/scala/slamdata/engine/javascript/javascript.scala [hunk 1]
```diff
@@ -108,7 +110,42 @@ object Js {
...
           List(arg))
     }
-  
+
+  def whenDefined(expr: Expr, body: Expr => Expr, default: => Expr): Expr =
+    expr match {
+      case Null   => Null
+      case _: Lit => body(expr)
+      case _      =>
+        Let(Map("expr" -> expr),
```
Comment (jdegoes) [inline core/src/main/scala/slamdata/engine/javascript/javascript.scala]: This has a simpler compilation that avoids a `Let`.

```javascript
typeof (<expr>) === "undefined" ? (<body>) : (<default>)
```

File: core/src/main/scala/slamdata/engine/javascript/javascript.scala [hunk 2]
```diff
@@ -108,7 +110,42 @@ object Js {
...
+  def safeCall(obj: Expr, fn: String, args: List[Expr]): Expr =
+    whenDefined(obj, obj => Call(Select(obj, fn), args), Null)
+
+  def safeDeref(expr: Expr, key: Expr): Expr =
+    whenDefined(
+      expr,
+      smartDeref(_, key),
+      // FIXME: this breaks in the rare case that someone adds an `__undef`
+      //        field to Object.
+      Select(AnonObjDecl(Nil), "__undef"))
```
Comment (jdegoes) [inline core/src/main/scala/slamdata/engine/javascript/javascript.scala]: Hack alert! :scream_cat:

File: core/src/main/scala/slamdata/engine/javascript/jscore.scala [hunk 3]
```diff
@@ -1,6 +1,6 @@
 package slamdata.engine.javascript
 
-import scala.collection.immutable.ListMap
+import scala.collection.immutable.Map
```
Comment (jdegoes) [inline core/src/main/scala/slamdata/engine/javascript/jscore.scala]: Curious why the change from `ListMap` to `Map`? Obviously does not affect semantics but will affect pretty printing, at least.

File: core/src/main/scala/slamdata/engine/physical/mongodb/bson.scala [hunk 4]
```diff
@@ -146,7 +146,7 @@ object Bson {
   case class JavaScript(value: Js) extends Bson {
     def bsonType = BsonType.JavaScript
 
-    def repr = new types.Code(value.render(2))
+    def repr = value.render(2)
```
Comment (jdegoes) [inline core/src/main/scala/slamdata/engine/physical/mongodb/bson.scala]: Has this approach been tested and found to work in all the cases in which MongoDB expects Javascript? Or does it fix some at the cost of breaking others???

File: core/src/test/scala/slamdata/engine/javascript/jscore.scala [hunk 5]
```diff
@@ -32,7 +32,11 @@ class JsCoreSpecs extends Specification with TreeMatchers {
       val b = JsMacro(JsCore.Select(_, "bar").fix)
```
Comment (jdegoes) [inline core/src/test/scala/slamdata/engine/javascript/jscore.scala]: Probably a test or two should be added to ensure the deref functionality is sufficiently robust (i.e. can handle non-trivial expressions in all the different parts). Also it looks like there are no tastes for the other safe functionality.

File: core/src/test/scala/slamdata/engine/physical/mongodb/planner.scala [hunk 6]
```diff
@@ -629,8 +627,13 @@ class PlannerSpec extends Specification with ScalaCheck with CompilerHelpers wit
           $map($Map.mapMap("__arg0",
             Call(AnonFunDecl(List("rez"),
               List(
-                ForIn(Ident("attr"),Select(Ident("__arg0"), "__tmp1"),If(Call(Select(Select(Ident("__arg0"), "__tmp1"), "hasOwnProperty"),List(Ident("attr"))),BinOp("=",Access(Ident("rez"),Ident("attr")),Access(Select(Ident("__arg0"), "__tmp1"),Ident("attr"))),None)),
-                BinOp("=",Access(Ident("rez"),Str("__sd__0")),Select(Ident("__arg0"), "__sd__0")), Return(Ident("rez")))),
+                ForIn(Ident("attr"), safeDeref(Ident("__arg0"), Str("__tmp1")),
+                  If(safeCall(safeDeref(Ident("__arg0"), Str("__tmp1")), "hasOwnProperty", List(Ident("attr"))),
```
Comment (jdegoes) [inline core/src/test/scala/slamdata/engine/physical/mongodb/planner.scala]: Makes me yearn for a Scala DSL for this stuff. :cry:

File: core/src/main/scala/slamdata/engine/javascript/javascript.scala (hunk 1 above)
Comment (sellout) [inline core/src/main/scala/slamdata/engine/javascript/javascript.scala]: It doesn’t, though – `body` takes `expr` as a parameter, and we need to avoid evaluating it twice.

File: core/src/test/scala/slamdata/engine/physical/mongodb/planner.scala (hunk 6 above)
Comment (mossprescott) [inline core/src/test/scala/slamdata/engine/physical/mongodb/planner.scala]: Indeed.

These are going to be a bitch to debug when they fail, since the values being compared don't look like the test source. Do we fix that by migrating this code to use JsCore/SlamScript, or is that not feasible for these ops?

**Comment by mossprescott** [comment]
Surprised to see only one new integration test/skeptical that that one test can cover the new behavior. What about partially-undefined values (selecting an undefined field from a sub-object that is present, etc.)

Also there is a pending test under `it/src/test/resources/tests/null` that may be interesting.

File: core/src/test/scala/slamdata/engine/physical/mongodb/planner.scala (hunk 6 above)
Comment (sellout) [inline core/src/test/scala/slamdata/engine/physical/mongodb/planner.scala]: I think we will need like a `Compilable[Target]` trait, so that `$map`, etc. can take anything that implements `Compilable[JavaScript]` and render whatever we gave it, rather than having to either convert everything to JS first or force Workflow to know about SlamScript.

File: core/src/main/scala/slamdata/engine/physical/mongodb/bson.scala (hunk 4 above)
Comment (sellout) [inline core/src/main/scala/slamdata/engine/physical/mongodb/bson.scala]: Ok, I’m going to revert this. I do think it works in all cases (as far as the MongoDB driver is concerned), but technically we _should_ be wrapping code in `types.Code`, but the problem is that `Code.toString` doesn’t do the right thing.

I want to fix our Bson to serialize better, but to do that right, I want to fixplate it, and that’s enough for its own PR.

**Comment by sellout** [comment]
This also fixes #531, now that JsExprPhase has been converted to JsCore.
